src/tsickle.ts

```diff
@@ -759,43 +797,87 @@ class Annotator extends ClosureRewriter {
+          symbols = [typeChecker.getSymbolAtLocation(importClause.name)];
         } else {
-          // import {a as b} from ...;
-          const namedImports = importClause.namedBindings as ts.NamedImports;
-          localNames = namedImports.elements.map(imp => getIdentifierText(imp.name));
-        }
-
-        for (let name of localNames) {
-          // This may look like a self-reference but TypeScript will rename the
-          // right-hand side!
-          this.emit(
-              `\nconst ${name}: NeverTypeCheckMe = ${name};  /* local alias for Closure JSDoc */`);
+          if (!importClause.namedBindings ||
+              importClause.namedBindings.kind !== ts.SyntaxKind.NamedImports) {
+            throw new Error('unreached');  // Guaranteed by if check above.
+          }
+          symbols =
+              importClause.namedBindings.elements.map(e => typeChecker.getSymbolAtLocation(e.name));
         }
+        this.forwardDeclare(decl.moduleSpecifier, symbols, !!importClause.name);
       }
       return true;
     } else if (
         importClause.namedBindings &&
         importClause.namedBindings.kind === ts.SyntaxKind.NamespaceImport) {
       // import * as foo from ...;
       this.visit(importClause);
-      this.emit(' from');
-      this.writeModuleSpecifier(decl.moduleSpecifier);
-      this.emit(';');
+      this.emit(` from '${importPath}';`);
       return true;
     } else {
       this.errorUnimplementedKind(decl, 'unexpected kind of import');
       return false;  // Use default processing.
     }
   }
 
+  private forwardDeclareCounter = 0;
+
+  /**
+   * Emits a `goog.forwardDeclare` alias for each symbol from the given list.
+   * @param specifier the import specifier, i.e. module path ("from '...'").
+   */
+  private forwardDeclare(specifier: ts.Expression, symbols: ts.Symbol[], isDefaultImport = false) {
+    if (this.options.untyped) return;
+    const importPath = this.resolveModuleSpecifier(specifier);
+    const nsImport = extractGoogNamespaceImport(importPath);
+    const forwardDeclarePrefix = `tsickle_forward_declare_${++this.forwardDeclareCounter}`;
+    const moduleNamespace =
+        nsImport !== null ? nsImport : this.pathToModuleName(this.file.fileName, importPath);
+    const typeChecker = this.program.getTypeChecker();
+    const exports = typeChecker.getExportsOfModule(typeChecker.getSymbolAtLocation(specifier));
+    // In TypeScript, importing a module for use in a type annotation does not cause a runtime load.
+    // In Closure Compiler, goog.require'ing a module causes a runtime load, so emitting requires
+    // here would cause a change in load order, which is observable (and can lead to errors).
+    // Instead, goog.forwardDeclare types, which allows using them in type annotations without
+    // causing a load. See below for the exception to the rule.
+    this.emit(`\nconst ${forwardDeclarePrefix} = goog.forwardDeclare('${moduleNamespace}');`);
```

Comment:
These emits are weird, why have the newline first?  Usually the tsickle pattern is one \n to start a block of emits, and then each emit has its own trailing \n